### PR TITLE
Support for JSON logging

### PR DESCRIPTION
**Pull request by gulshan-rs**
### All Submissions:

Enhances https://github.com/qdrant/qdrant/issues/5347 

I haven't been able to test this by just running the integration tests, so please suggest me how to test this.

### CODE REVIEW

**Comment by gulshan-rs** [comment]
cc @timvisee if you get a chance to look at this one :) 

ps: I think the integ test failure is not related, not sure how do i trigger a new run without doing a force push.

File: src/tracing/default.rs [hunk 1]
```diff
@@ -2,17 +2,26 @@ use std::collections::HashSet;
...
+use tracing_subscriber::fmt::{self};
+use tracing_subscriber::{filter, registry, Layer};
 
 use super::*;
 
+#[derive(Default, Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum LogFormat {
+    #[default]
+    Default,
```
Comment (timvisee) [inline src/tracing/default.rs:14]: Maybe we can name the default `Text`, or something like that. I'm actually not sure what I like best :smiley: 

```suggestion
    #[default]
    Text,
```

File: src/tracing/default.rs [hunk 2]
```diff
@@ -21,39 +30,56 @@ impl Config {
...
+    filter::Filtered<Option<Box<dyn Layer<S> + Send + Sync>>, filter::EnvFilter, S>;
 
 pub fn new_logger<S>(config: &Config) -> Logger<S>
 where
     S: tracing::Subscriber + for<'span> registry::LookupSpan<'span>,
 {
-    let layer = new_layer(config);
+    let layer: Box<dyn Layer<S> + Send + Sync> = match config.log_format {
+        Some(LogFormat::Json) => Box::new(new_layer_with_json(config)),
+        _ => Box::new(new_layer(config)),
```
Comment (timvisee) [inline src/tracing/default.rs:52]: I'd prefer explicit matching here:

```suggestion
        Some(LogFormat::Json) => Box::new(new_layer_with_json(config)),
        Some(LogFormat::Default) | None => Box::new(new_layer(config)),
```

Not super critical here, but probably something nice to get used to.

Comment (gulshan-rs) [reply src/tracing/default.rs:52]: agreed, its more readable.

File: src/tracing/default.rs (hunk 1 above)
Comment (gulshan-rs) [reply src/tracing/default.rs:14]: I like it being ```Text```

File: src/tracing/default.rs [hunk 3]
```diff
@@ -2,17 +2,26 @@ use std::collections::HashSet;
...
 use serde::{Deserialize, Serialize};
-use tracing_subscriber::prelude::*;
-use tracing_subscriber::{filter, fmt, registry};
+use tracing_subscriber::fmt::{self};
+use tracing_subscriber::{filter, registry, Layer};
 
 use super::*;
 
+#[derive(Default, Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
```
Comment (timvisee) [inline src/tracing/default.rs:11]: Nitpick: not necessary now, but probably a better default if we ever add more variants here.

```suggestion
#[serde(rename_all = "snake_case")]
```

**Review by timvisee — APPROVED** [review]
Thanks! Works like a charm :pray:

**Comment by gulshan-rs** [comment]
think we can close https://github.com/qdrant/qdrant/issues/5347

**Pull request merged**